src/lib.rs

```diff
@@ -210,88 +132,327 @@ fn get_repo(pkg_info: &Value) -> Result<Option<String>, Error> {
+        Ok(Cloner {
+            registry_url: DEFAULT_REGISTRY_URL.to_string(),
+            github_url: DEFAULT_GITHUB_URL.to_string(),
+            gitlab_url: DEFAULT_GITLAB_URL.to_string(),
+            bitbutcket_url: DEFAULT_BITBUCKET_URL.to_string(),
+            out_dir: env::current_dir()?,
+        })
+    }
 
-        // Sanity check.
-        if !entry_path.starts_with(&base) {
+    /// Clones a crate using the provided method.
+    ///
+    /// - `method_kind` - Method to fetch crate.
+    /// - `spec` - The name of the crate to clone
+    /// - `version` - The semantic version (semver) of the spec crate to clone
+    /// - `extra` - Additional arguments passed to clone command.
+    ///
+    pub fn clone(
+        &self,
+        method_kind: CloneMethodKind,
+        spec: &str,
+        version: Option<&str>,
+        extra: &[&str],
+    ) -> Result<(), Error> {
+        let mut parts = spec.splitn(2, ':');
+        let name = parts.next().unwrap();
+        let spec_version_req = parts.next();
+        if spec_version_req.is_some() && version.is_some() {
+            bail!("Cannot specify both a :version and --version.");
+        }
+        let version_req = version
+            .or(spec_version_req)
+            .map(check_semver_req)
+            .transpose()?;
+        let pkg_info = self.get_pkg_info(name)?;
+        let repo = get_repo(&pkg_info)?;
+        let (method, repo) = match method_kind {
+            CloneMethodKind::Auto => {
+                if version_req.is_some() {
+                    (CloneMethodKind::Crate, "".to_string())
+                } else if let Some(repo) = repo {
+                    self.detect_repo(&repo)?
+                } else {
+                    (CloneMethodKind::Crate, "".to_string())
+                }
+            }
+            CloneMethodKind::Crate => (method_kind, "".to_string()),
+            _ => {
+                if repo.is_none() {
+                    bail!("Could not find repository path in crates.io.");
+                }
+                (method_kind, repo.unwrap())
+            }
+        };
+        match method {
+            CloneMethodKind::Crate => {
+                if !extra.is_empty() {
+                    bail!("Got extra arguments, crate downloads take no extra arguments.");
+                }
+                self.clone_crate(name, version_req, &pkg_info)?;
+            }
+            CloneMethodKind::Git
+            | CloneMethodKind::Mercurial
+            | CloneMethodKind::Pijul
+            | CloneMethodKind::Fossil => {
+                if let Some(version_req) = version_req {
+                    bail!(
+                        "Specifying a version `{}` only works with the `crate` method.",
+                        version_req
+                    );
+                }
+                self.run_clone(method.command(), &repo, extra)?;
+            }
+            _ => bail!("Unsupported method `{}`", method),
```

Comment:
```suggestion
            CloneMethodKind::Auto => unreachable!(),
```

Reply:
Applied this change manually, I didn't know about `unreachable`, pretty cool! 😄